File: test/cpp/src/fake_camera_display.cpp
/*

g++ src/fake_camera_display.cpp -o fake_camera_display  \
-I /opt/sensing-dev/include -I /opt/sensing-dev/include/aravis-0.8 \
-L /opt/sensing-dev/lib \
-L /opt/sensing-dev/lib/x86_64-linux-gnu \
-lHalide -lion-core -ldl -lpthread \
-laravis-0.8 -lgobject-2.0 \
`pkg-config --cflags --libs glib-2.0`

*/

#include <ion/ion.h>

#include <exception>
#include <string>
#include <map>

using namespace ion;

std::map<std::string, std::string> bb_name{
  {"Mono8", "image_io_u3v_cameraN_u8x2"},
  {"Mono10", "image_io_u3v_cameraN_u16x2"},
  {"Mono12", "image_io_u3v_cameraN_u16x2"}};

int main(int argc, char *argv[])
{
  try{
    const int32_t width = 128;
    const int32_t height = 128;
    int32_t num_device = 2;
    std::string pixelformat = "Mono8";


    Builder b;
    b.set_target(ion::get_host_target());
    b.with_bb_module("ion-bb");

    // add BB to pipeline
    Node n = b.add(bb_name[pixelformat])()
      .set_params(
        Param("num_devices", num_device),
        Param("force_sim_mode", true),
        Param("width", width),
        Param("height", height)
      );

    // portmapping from output port to output buffer
    std::vector< int > buf_size = std::vector < int >{ width, height };

    std::vector<Halide::Buffer<uint8_t>> output;
    for (int i = 0; i < num_device; ++i){
      output.push_back(Halide::Buffer<uint8_t>(buf_size));
    }
    n["output"].bind(output);

    for (int i = 0;i<10;i++)
    {
      // JIT compilation and execution of pipelines with Builder.
      b.run();
      // Convert the retrieved buffer object to OpenCV buffer format.
    }

    for (int i = 0; i < num_device; ++i){
        uint8_t img[height][width];
        std::memcpy(img, output[i].data(), output[i].size_in_bytes());

        for(int row = 0; row < height; row++)
        {
          for(int col = 0; col < width; col++){
               std::cout << unsigned(img[row][col]) << " ";
          }
          std::cout << std::endl;
        }
    }
  } catch(std::exception& e){
    std::cerr << e.what() << std::endl;
    return 1;
  } catch (const ion::Error& e) {
    std::cerr << e.what() << std::endl;
    return 1;
  }
  return 0;
}

File: test/python/fake_camera_display.py
import numpy as np
from ionpy import Node, Builder, Buffer, Port, Param, Type, TypeCode

if __name__ == "__main__":

    # device information
    width = 640
    height = 480
    pixelformat = "Mono8"
    num_device = 2

    # the following items varies by PixelFormat
    data_type = np.uint8 if pixelformat == "Mono8" \
        else np.uint16 if pixelformat == "Mono16" \
        else np.uint8
    depth_of_buffer = np.iinfo(data_type).bits
    depth_in_byte = int(depth_of_buffer / 8)
    bb_name = "image_io_u3v_cameraN_u8x2" if pixelformat == "Mono8" \
        else "image_io_u3v_cameraN_u16x2" if pixelformat == "Mono16" \
        else "image_io_u3v_cameraN_u8x2"

    # pipeline setup
    builder = Builder()
    builder.set_target('host')
    builder.with_bb_module('ion-bb')

    # set params
    num_devices = Param('num_devices', num_device)
    frame_sync = Param('frame_sync', True)
    realtime_display_mode = Param('realtime_display_mode', True)

    # add a node to pipeline
    node = builder.add(bb_name) \
        .set_params([num_devices, frame_sync, realtime_display_mode,
                     Param('width', width),
                     Param('height', height)])
    output_p = node.get_port('output')

    # create halide buffer for output port
    outputs = []
    output_datas = []
    output_size = (height, width,)

    for i in range(num_device):
        output_datas.append(np.full(output_size, fill_value=0, dtype=data_type))
        outputs.append(Buffer(array=output_datas[i]))

    # set I/O ports
    output_p.bind(outputs)

    for i in range(10):
        # running the builder
        builder.run()
        print(output_datas[0])
        print(output_datas[1])



